### ISSUE TITLE
TabularDataset class with header does not match fields by name

### ISSUE THREAD

**Issue by keitakurita** [pytorch/text#215]
Currently, the TabularDataset class matches each feature in a csv/tsv file to its field by requiring the (name, Field) tuples passed to the fields kwarg to be in the same order as the columns csv/tsv file.

I think this is counterintuitive, especially if the csv/tsv file has a header. If there is a header, I think that the user should be allowed to specify the fields similar to the json format, where they pass in a dict mapping the column name to the (name, Field) tuple.

Here’s an example of what I’m thinking.

```
$ head -n 2 train.csv
"text","label"
"hello",0
```

```python
>>> pos = data.TabularDataset(
...    path='train.tsv', format='csv',
...    fields={'labels': ('labels', data.Field()),
...                'text': ('text', data.Field())}
...
```

I think they should also be allowed to select a subset of the rows using this method, similar to the API for the json format (columns not in the keys will be ignored).

This should provide a more consistent and intuitive API. If the maintainers are willing to accept this API, I would be happy to implement it and send a PR.

**Comment by jekbradbury** [comment]
Yes, we'd welcome a PR as long as it doesn't break the current API for headerless files.

**Comment by jekbradbury** [comment]
Closed by #217
